monai/data/dataset.py

```diff
@@ -66,19 +68,24 @@ def __init__(self, data: Sequence, transform: Optional[Callable] = None) -> None
     def __len__(self) -> int:
         return len(self.data)
 
-    def __getitem__(self, index: int):
-        data = self.data[index]
-        if self.transform is not None:
-            data = apply_transform(self.transform, data)
+    def _transform(self, index: int):
+        return apply_transform(self.transform, self.data[index]) if self.transform is not None else self.data[index]
 
-        return data
+    def __getitem__(self, index: int):
+        if isinstance(index, slice):
+            start, stop, step = index.indices(len(self))
+            index_list = list(range(start, stop, step))
+            return [self._transform(i) for i in index_list]
```

Comment:
I thought about the same but logically this should return a subset dataset `Dataset(data=self.data[slice], transform=transform)`. what would be the expected behaviour for the cached version? now I'm not sure... cc @rijobro  @ericspod

Reply:
@rijobro @ericspod , could you please help share some comments here?

Thanks in advance.

Reply:
We should use [Subset](https://pytorch.org/docs/stable/data.html#torch.utils.data.Subset) for getting subsets of existing datasets, I don't think we should necessarily support slice in this way.

Reply:
Hi @ericspod ,

Sounds good, let me update the PR to return a `Subset` instead.
Thanks.